make worker threads robust and able to catch catastrophic failures for more controlled general application shutdown and post mortem.

GerHobbelt opened the pull request:
> **Note**
>
> This work builds upon https://github.com/bshoshany/thread-pool/pull/76, hence it is merged as part of this commit chain to the individual commit deltas small, comprehensible and mergeable.

Adding *virtual member function* `workerthread_main()` which is used to wrap every worker thread's `worker()` method: this prepares us for two features to be submitted:

1. per thread catastrophic failure handling, where worker threads gain the ability to detect catastrophic failures, e.g. uncaught exceptions and hardware faults, and help to terminate the application in a controlled fashion.

2. allowing applications to use a derived class which is used to provide advanced per-worker-thread setup and shutdown procedures which are application-specific. (Hence the *virtual* method.)

---

Merge branch 'patch-7' (i.e. https://github.com/bshoshany/thread-pool/pull/76 ) -- patch-7 is a mandatory prerequisite for this work.

---


added the `get_alive_threads_count()` monitoring feature, which tracks the worker threads 'life': this is in preparation of the catastrophic failure handling feature.

---


## Revision: 598f66ff8f5e183c12f378daca0593f71d43c7ff

Phase 1 of the per-workerthread catastrophic failure handling code: this contains the platform-generic part, which is able to catch unhandled exceptions and report those to any custom `workerthread_main()` reporter.

> This is the inverse of commit ad63b02290e4debeeae388bbce925426cb023071 as it was easier to remove the features one-by-one when I was extracting this bit of work from the dev commits, then revert those onto here.

Originally From SHA-1: 12253b6610ed2f237660abeb91446beab27f63d1

**augmentations**:

- thread workers now wait for a signal to wake up when in 'pause mode', as they always already were in normal run mode.

[...]

  Note that, at least under Windows 10, threads MAY be nuked silently under fatal/crash/exception conditions, in which case the thread disappears even before it was able to do anything simple, like updating the `alive_threads_count` count, because the thread worker code simply will cease to run and exist. Fortunately, thread::is_joinable() detects this fact -- which was a very important reason to have `wait_for_tasks()` be a wait_for_tasks()-timed poll loop, as we CANNOT guarantee on all OSes that worker threads will be able to wake up and act once a severe-enough fault condition has occurred in the application/task code. From our own observations, it already seems sufficient to directly, bluntly, call the standard `exit(N)` RTL API to have this happen to you: no exception of any kind will be reported then, yet all threads will be joinable as they will have *vanished* already.

WARNING: hence, `get_alive_threads_count()` will be unsafely optimistic and depending on that number at such disparate times will surely cause your application to lock up on exit on semi-random occasions. This is completely independent of the `running` state, as this is driven by external factors.

- catch unexpected C++ and SEH/hardware exceptions occurring in your tasks/worker-threads in the outer layers of the worker thread: as this is a catastrophic, fatal, condition anyway (your application state is largely unpredictable by then already), the thread will terminate, but we now have a fighting chance of catching and reporting such errors at least. As C++ and SEH exception handling cannot co-exist in a single function, we have the following call chain, where each wraps the next one: workerthread_main() --> __worker_SEH() --> __worker() --> worker(), where worker() is the core threadpool thread code, waiting for and executing tasks once they arrive.

  When you need special handling of this (fatal) scenario in your application, you can create a derived class and override the workerthread_main() with your own. Observe the code comments when you do to ensure continued proper operation of the threadpool. (`__worker_SEH()` returns a boolean indicating whether its termination was due to normal or abnormal (i.e. catastrophic failure) termination, while you may pass a string to them, which will now have been filled with the relevant and available error information. The given implementation prints this info the STDERR -- but you can replace that behaviour in your override.

----



## Revision: 57d067a405c646f8f1e43f1c949b2c57bb7b7b66

Add support for *negative* `thread_count`: '*this few less than number of CPU cores*' setup feature.

From the original: SHA-1: 12253b6610ed2f237660abeb91446beab27f63d1

- now accepts NEGATIVE `thread_counts` during pool creation (or `reset()`) to create a pool that's occupying *all CPU cores MINUS thread_count*. This is useful when you want the pool to occupy ALMOST all cores, but leave one or more for other, unrelated, jobs.

**... Which is a feature due to the results from SHA-1: de3247306c4ae9e370e740579deb6e235ef96b9a: see https://github.com/bshoshany/thread-pool/pull/74 and https://github.com/bshoshany/thread-pool/pull/75 for that bit of the work.**

> This is produced as a revert of commit 3109b5c146f1ce50855b92ba57a8324a160588b3 as it was easier to remove features from a existing and tested-ok rig than build up from scratch on a new tree.





## Revision: 635472fff414f6b3d82fa98aaeb7147ddc70ba02

`wait_for_tasks()`: remove the need to use YIELD by slowly growing the poll period until it hits a preconfigured ceiling (1 second); this code is generally only relevant in abnormal application terminations, but helps to keep it civil at those times.

> This reverts commit f7551e41cd5c85d724dfb47b3d94ec03afc76798 as it was easier to remove features from an existing and known-good code rig than build up from scratch on a different tree, so produced a remove patch there and then reverted it onto here. *Ach.*🤷‍♀️



## Revision: 34b6aa53e3453ea39209e6510b945f16b9e10dd7

add the original `ASSERT`ion checks, which document and test the implicit assumptions at critical places in the code. We *explicitly decide* NOT to use STD RTL `<assert.h>` et al as those are very inflexible, not allowing one, for example, to continue debugging an application when an assertion fails. These are engineered to be replaced by application/test specific macros; their default implementation (on Windows at least) will wake up your attached debugger -- if you were running any.

> Works nicely for me at least with a MSVC2019 dev rig. 😄 

> Again, this is another inverse of a commit: 72424f2c65e1b2cbe36c781b5dc35971b66d5417. It was easier to remove features from a known-good rig and then revert those commits onto here.🤷‍♀️




## Revision: 97b10acc32ac21b50089a31e7a1c5b919f3add59

Phase 2 of the per-threadworker catastrophic failure handling and controlled termination: adding [MSWindows-specific SEH handling](https://docs.microsoft.com/en-us/cpp/cpp/structured-exception-handling-c-cpp?view=msvc-170), which allows an application to catch and process hardware failures of various kinds, e.g. bus errors and DIV/0. This allows advanced applications to cope with internal failures and shut down in a controlled fashion - even while the state of a large chunk of the application may arguably already be 'in limbo'.

"adding the MSVC-specific extra SEH catastrophic failure handling code. Based on the already-present generic skeleton rig, so the change is minimal. This basically fleshes out the `__worker_SEH()` wrapper method."

> This reverts commit 4c0774a860a1534aae19424893dfe619b21fcac8. It was easier to remove features from a known-good rig than to build up from scratch on a different tree, so we created the removal commit there, then reverted that one onto here.




## Revision: bb550265245caae76e137531baf1b8527c85265b

Continuation of SHA-1: 59c4193a44d9bdb858a1d66ca90ea9c62bf1331a, adding the omitted original `ASSERT`ion checks, which are applicable now that the SEH handling code in `__worker_SEH()` has finally been introduced.


---



🥴😱 whoopsie! forgot to add the `get_alive_threads_count()` test which was added to check the new added API. Done now. ✅





---

**Testing**

This was tested as part of a larger work (see the previous PRs) after hunting down shutdown issues (application lockups, etc.) in a large application.

Tested this code via your provided test code rig; see my own fork and the referenced commits which point into there.

Tested on AMD Ryzen 3700X, 128GB RAM, latest Win10/64, latest MSVC2019 dev environment. Using in-house project files which use a (in-house) standardized set of optimizations. 

**Additional information**

**TBD**

The patches are hopefully largely self-explanatory. Where deemed useful, the original commit messages from the dev fork have been referenced and included.

bshoshany commented:
Thanks for your 9 (!) pull requests #70-#78. I really appreciate your contribution!

But I think you've been patching an old version of the library, since you seem to be using the `sleep_duration` variable, which was eliminated in v3.0.0. Therefore some or all of your changes may no longer be relevant. But I will check your suggested changes and may incorporate some in the next release if they seem relevant to the latest version.

Unfortunately I'm a bit busy over the next few weeks, and there are a lot of changes to go over, so it might take a while before I am able to respond to each of your pull requests.

GerHobbelt commented:
I re-introduced that variable for wait_for_tasks() as that one needed a
poll loop + .wait_for() instead of simple .wait().

So yeah at first glance it might look like a patch against V2, but it sure
is developed and tested against latest v3 (v3.3.0 if my hit reads right).


On Sun, Aug 21, 2022, 23:27 Barak Shoshany ***@***.***> wrote:

> Thanks for your 9 (!) pull requests #70
> <https://github.com/bshoshany/thread-pool/pull/70>-#78
> <https://github.com/bshoshany/thread-pool/pull/78>. I really appreciate
> your contribution!
>
> But I think you've been patching an old version of the library, since you
> seem to be using the sleep_duration variable, which was eliminated in
> v3.0.0. Therefore some or all of your changes may no longer be relevant.
> But I will check your suggested changes and may incorporate some in the
> next release if they seem relevant to the latest version.
>
> Unfortunately I'm a bit busy over the next few weeks, and there are a lot
> of changes to go over, so it might take a while before I am able to respond
> to each of your pull requests.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bshoshany/thread-pool/pull/78#issuecomment-1221624003>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADCIHVNUPHIUETVYPBDMMLV2KNMPANCNFSM57FQZO3A>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

bshoshany commented:
Thanks for the explanation. This behavior was eliminated in v3.0.0 because it caused high CPU usage, decreased performance, and forced the user to choose a value for the `sleep_duration` parameter by hand. Removing it was an often-requested feature. Therefore I cannot accept this pull request, as it reintroduces unwanted behavior. In addition, your pull request introduces a lot of additional complexity and I try to keep the code as lightweight and simple as possible. However, I do appreciate your contribution.

GerHobbelt commented:
Ok.

~~Before you kill https://github.com/bshoshany/thread-pool/pull/76 as "at first glance" it uses this `sleep_duration` variable: I did not explain it well, apparently. The **name** `sleep_duration` is re-used, but its purpose and activity is empathetically different from what is was in v2! I could have named it entirely different and we wouldn't be in this "this looks like v2 popping back up" confusion. 😭~~ 

~~## What's it for then?~~

~~It is only required to ensure the application / threadpool can terminate (destroy) itself **without locking up waiting forever in `thread.join()` under *all* circumstances, both normal and abnormal.~~

~~For regular operations, it is not important and thus can be "slow": it needs to be there because you fundamentally don't know when you'll hit the application termination phase in the application lifetime, so there's a balance point here. As far as I'm concerned it doesn't even need to configurable when push comes to shove: you can set it to several **seconds** and it'll still deliver, because all it says is: "when the application terminates and you encounter the situation where a thread has either already died or has not been observant for a while and thus was too late into its own `.wait()` for your last `notify_all()` to wake it up and **observe** your `running` state change, you are guaranteed to have, in the latter scenario, a second, third, etc chance to `.notify` and thus wake up every *late* thread once it has arrived at its own `.wait()` and thus have it terminate properly as per design, while you are thus guaranteed, in *both* scenarios, to have the controlling (destroying) thread/process not lock up because `destroy_threadpool()` will now ensure all (even the extremely tardy) threads will terminate *iff* they haven't already and it will not depend on threads signalling back as it uses the `thread.is_joinable()` standard API to check when, finally, all threads have died and thus termination of all has ended and the process can then proceed to terminate itself, in an controlled and orderly fashion.~~

[*Removed explanation of intent, problem and path towards & selection of solution.*]

While writing, it also made me rethink the entire thing, once more. Turns out I initially ran into issues regarding the guarantees of the "modification order" (as described in these pages: https://en.cppreference.com/w/cpp/thread/condition_variable/wait) thanks to other code corruption (invalid pointers doing nasty stuff thanks to heap corruption and its after-effects), then, while debugging that, discovering it's essentially using a subtly **not-threadsafe** library in a multithreaded setting, ran into the continued lock-up at application termination, particularly when running the application within the (MSVC) debugger.

Your response drove me to reconsider https://github.com/bshoshany/thread-pool/pull/76 and ask myself why it still feels worthwhile, while my re-investigation says it can be done without **any** `.wait_for()` and still deliver its controlled exit guarantees.

Thank you. 

To quote a certain someone: "I'll be back!" 😉

bshoshany commented:
I see, thanks for the explanation! So to confirm, are you saying the problem you encountered in your large application is not due to a bug in my library, but due to another unrelated library?

GerHobbelt commented:
Root cause: another lib.

Result: on my way to uncover the root cause, I had to take apart yours. Combination of debugging/observation + code review because the problem wasn't deterministic.

I'm staying away for now from saying your library wasn't part of the bug as I really need to rethink all scenarios through before I can say something I believe in again. (Back to the drawing board sort of thing. :wink:)

Why I am not confident enough to say v3.3.0 is clean?

The road traveled was... complicated.

In fact, I started out with a copy of thread_pool v2, when the application went belly-up. While I was trying to get to the root cause, where I only had a *suspicion* of which library was making a mess, but not any tighter target zone than that, thread_pool popped up on the bug radar as v2 went completely nuts in the debugger: a repeat performance which resulted in me taking it apart, doing a thorough code review (while I knew it wasn't the prime suspect but it wasn't helping matters and while the other problem driver stayed elusive, this was a doable angle of attack. Turned out the crap caused elsewhere triggered the thread_pool v2 code into looping like mad (IIRC it was `wait_for_task()` then doing the nasty; anyway, one of the `.wait()`s was effectively short-circuiting. So I fixed it for v2, but code review had me thinking there was more subtle failure scenarios, so I looked up your origin repository and took latest v3 to "upgrade" the application to that one as it looked solid at first glance and could only matters as it was.

While I was completely stumped about how dig out the root cause, I had a deeper look at v3 (code review) and took your sample applications and tested them: better safe then sorry and all that.

v3 read quite different than v2, but there's one part that popped up early in the v3.3.0 code review that I now recall as I read the #76 pullreq patch again: when in pause mode, the `wait()` condition check is not complete as you can come up with scenarios where you have a `running=true` plus `tasks.empty() == false` so **exit from .wait()`** and then you're running the worker thread loop like a madman for `.wait()` will be short-circuited for the duration. While that came out of code review, I **observed** thread_pool going wild again (this time in a worker instead of v2 which, IIRC, went nuts in `wait_for_tasks()`), while the other lib was making lots of trouble by corrupting tasks, so the entire bloody thing was a moving target (and you'd see a crash in a few spots, but every re-run was running a lottery about where it would b0rk *this time* 👿   

So I trusted my code review instincts, took thread_pool and 'ported' my changes from my v2 copy into v3, full well realizing the flow was entirely different, so I did my best to check my reasoning and picking only the useful bits. Because v3 was reasonably reliably misbehaving, also in your test code, I had at least a situation which was more or less similar to the big one while the codebase was much smaller, so that was one long session. Next day, things were stabilized and predictable enough that root cause could be narrowed down. A hunch led to putting a critical section around all the spots the other library was invoked and the semi-random memory corruptions stopped entirely. Meanwhile every bit that was considered a risk, even if was deemed infinitesimally small, that I could find in your code, was fixed. Then I introduced forced bus errors into the app codebase (invalid addresses, etc.) to analyze the failure paths again, because I was pretty sure I wouldn't find the **exact** root cause and fix it in that lib, so second best is catching it as soon as it happens.

That's where `.wait_for()` for `wait_for_tasks()` became a necessity, after checking the documentation, so that mandated `sleep_duration`.

Your replies and me wanting to provide a useful response led to tonight's discovery, which needs to be thought through for all error cases before I can be sure:

`wait_for_tasks()`, which I re-introduced in `destroy_tasks()`, is not necessary there (iff the C++ documentation and MSVC code is reliable, of course :wink:) if the `.wait()` in the thread_workers uses a correct, *complete* condition check. What IS needed is a `.notify_all()` in `destroy` because the already `.wait`ing worker threads need that trigger to re-evaluate the condition in order to observe `running==false` (while `paused==true` or `paused==false`) and terminate, making them joinable. Meanwhile, they MUST use `.notify_all()` themselves (instead of `.notify_one` as they do now), as there's the potential for multiple application threads calling `wait_for_tasks()` (for their own purposes) and being stuck in the `.wait()` therein: now imagine an application termination being triggered elsewhere, hence `running` changes to `running==false`: awareness of this needs to be seen by all waiting `wait_for_tasks()` as well as all thread_workers and there may or may not be any thread_workers() around to `.notify`-wake the N `wait_for_tasks()` (N >= 1) into checking this state in their own `.wait()` condition. (Yes, that's pretty convoluted perhaps, but once I got to think about multiple control threads calling `wait_for_tasks()` and seeing the potential trouble there, I could quickly reduce it down to a single application thread doing that and 'external forces' (i.e. another spot in the app) triggering the change to `running==false` and producing the same *lock-up at termination* problems, that I was observing in the debugger.

Now that I write this, I realize I'm *still* unsure about the exact flow under all circumstances, so, please, this needs a serious recheck. Because I feel I might have missed a corner case there, but don't know right now which one. Brain needs to come up with an "all clear!" or "here's your scenario!" about the perceived need for `.wait_for()` in `wait_for_tasks()` - sorry. It's not simple.

GerHobbelt commented:
(sorry, fixed typos in comment above. Too late, need to catch some Z's here.)

GerHobbelt commented:
Before I hit the bed, note to self:

One of the driving reasons to pick `.wait_for()` was the documented equivalence to 'condition check **before** .wait() is reached' (forget the loop bit there, it's about the moment the execution path *entering* this line of code) and me (I now believe) missing that `.wait(..., condition)` delivers the same (as per https://en.cppreference.com/w/cpp/thread/condition_variable/wait, the '2)' section in there while I read that documentation too.  
What makes this a "have-to-check" instead of "gotcha / facepalm" is the undesirable waiting-forever behaviour I got in `destroy()` when introducing hardware faults in the external code. Of course, I nuked that dirty test hack 😭 so might need to redo if rethink/review is inconclusive / not rock solid.

bshoshany commented:
Thanks for all the information! Like I said in the other PR, if you believe there's a bug in the library, please provide a minimal working example that reproduces the error. Looking forward to it.
